sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/OnTimerInvokers.java

```diff
@@ -0,0 +1,270 @@
+import net.bytebuddy.description.modifier.Visibility;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.dynamic.DynamicType;
+import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
+import net.bytebuddy.dynamic.scaffold.InstrumentedType;
+import net.bytebuddy.dynamic.scaffold.subclass.ConstructorStrategy;
+import net.bytebuddy.implementation.Implementation;
+import net.bytebuddy.implementation.bytecode.ByteCodeAppender;
+import net.bytebuddy.implementation.bytecode.StackManipulation;
+import net.bytebuddy.implementation.bytecode.member.FieldAccess;
+import net.bytebuddy.implementation.bytecode.member.MethodInvocation;
+import net.bytebuddy.implementation.bytecode.member.MethodReturn;
+import net.bytebuddy.implementation.bytecode.member.MethodVariableAccess;
+import net.bytebuddy.jar.asm.MethodVisitor;
+import net.bytebuddy.matcher.ElementMatchers;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.DoFn.TimerId;
+import org.apache.beam.sdk.transforms.reflect.DoFnInvokers.DoFnMethodDelegation;
+
+/**
+ * Dynamically generates {@link OnTimerInvoker} instances for invoking a particular {@link
+ * TimerId} on a particular {@link DoFn}.
+ */
+class OnTimerInvokers {
+  public static final OnTimerInvokers INSTANCE = new OnTimerInvokers();
+
+  /**
+   * The field name for the delegate of {@link DoFn} subclass that a bytebuddy invoker will call.
+   */
+  private static final String FN_DELEGATE_FIELD_NAME = "delegate";
+
+  /**
+   * A cache of constructors of generated {@link OnTimerInvoker} classes, keyed by {@link DoFn}
+   * class and then by {@link TimerId}.
+   *
+   * <p>Needed because generating an invoker class is expensive, and to avoid generating an
+   * excessive number of classes consuming PermGen memory in Java's that still have PermGen.
+   */
+  private final Map<Class<?>, Map<String, Constructor<?>>> constructorCache = new LinkedHashMap<>();
+
+  private OnTimerInvokers() {}
+
+  /** Creates invoker. */
+  public <InputT, OutputT> OnTimerInvoker invokerForTimer(
+      DoFn<InputT, OutputT> fn, String timerId) {
+
+    DoFnSignature signature = DoFnSignatures.INSTANCE.getSignature(fn.getClass());
+
+    try {
+      @SuppressWarnings("unchecked")
+      OnTimerInvoker<InputT, OutputT> invoker =
+          (OnTimerInvoker<InputT, OutputT>)
+              getByteBuddyOnTimerInvokerConstructor(signature, timerId).newInstance(fn);
+      return invoker;
+    } catch (InstantiationException
+        | IllegalAccessException
+        | IllegalArgumentException
+        | InvocationTargetException
+        | SecurityException e) {
+      throw new RuntimeException(
+          String.format(
+              "Unable to bind @%s invoker for %s",
+              DoFn.OnTimer.class.getSimpleName(), fn.getClass().getName()),
+          e);
+    }
+  }
+
+  @Nullable
+  private Constructor<?> getConstructorFor(Class<? extends DoFn<?, ?>> fnClass, String timerId) {
```

Comment:
The data flow between these two methods and getByteBuddyOnTimerInvokerConstructor seems confusing and I can't tell if they're doing anything redundant - it'd be easier to read IMO if they were inlined.